proxy/src/main/java/net/md_5/bungee/BungeeBossBar.java

```diff
@@ -0,0 +1,284 @@
+    {
+        this.style = Preconditions.checkNotNull( style, "style" );
+        if ( visible )
+        {
+            setDivisions( color, style );
+        }
+    }
+
+    @Override
+    public void setVisible(boolean visible)
+    {
+        boolean previous = this.visible;
+        if ( previous && !visible )
+        {
+            sendToAffected( removePacket() );
+        } else if ( !previous && visible )
+        {
+            sendToAffected( addPacket() );
+        }
+        this.visible = visible;
+    }
+
+    @Override
+    public Collection<BarFlag> getFlags()
+    {
+        return ImmutableList.copyOf( flags );
+    }
+
+    @Override
+    public void addFlags(BarFlag... flags)
+    {
+        if ( this.flags.addAll( Arrays.asList( flags ) ) && visible )
+        {
+            sendToAffected( updateFlags() );
+        }
+    }
+
+    @Override
+    public void removeFlag(BarFlag flag)
+    {
+        if ( flags.remove( flag ) && visible )
+        {
+            sendToAffected( updateFlags() );
+        }
+    }
+
+    @Override
+    public void removeFlags(BarFlag... flags)
+    {
+        if ( this.flags.removeAll( Arrays.asList( flags ) ) && visible )
+        {
+            sendToAffected( updateFlags() );
+        }
+    }
+
+    private byte serializeFlags()
+    {
+        byte flagMask = 0x0;
+        if ( flags.contains( BarFlag.DARKEN_SCREEN ) )
+        {
+            flagMask |= 0x1;
+        }
+        if ( flags.contains( BarFlag.PLAY_BOSS_MUSIC ) )
+        {
+            flagMask |= 0x2;
+        }
+        if ( flags.contains( BarFlag.CREATE_WORLD_FOG ) )
+        {
+            flagMask |= 0x4;
+        }
+        return flagMask;
+    }
+
+    private void setDivisions(BarColor color, BarStyle division)
+    {
+        BossBar packet = new BossBar( uuid, 4 );
+        packet.setColor( color.ordinal() );
+        packet.setDivision( division.ordinal() );
+        sendToAffected( packet );
+    }
+
+    private BossBar updateFlags()
+    {
+        BossBar packet = new BossBar( uuid, 5 );
+        packet.setFlags( serializeFlags() );
+        return packet;
+    }
+
+    private BossBar addPacket()
+    {
+        BossBar packet = new BossBar( uuid, 0 );
+        packet.setTitle( ComponentSerializer.toString( title ) );
+        packet.setColor( color.ordinal() );
+        packet.setDivision( style.ordinal() );
+        packet.setHealth( progress );
+        packet.setFlags( serializeFlags() );
+        return packet;
+    }
+
+    private void sendToAffected(DefinedPacket packet)
+    {
+        for ( ProxiedPlayer player : players )
+        {
+            sendPacket( player, packet );
+        }
+    }
+
+    private void sendPacket(ProxiedPlayer player, DefinedPacket packet)
+    {
+        if ( player.isConnected() )
+        {
+            player.unsafe().sendPacket( packet );
+        }
+    }
+
+    private BossBar removePacket()
+    {
+        return new BossBar( uuid, 1 );
```

Comment:
This packet could and should be lazily cached or actively created at object creation as final variable after `uuid` has been initialized, because it will always stay the same.